### ISSUE TITLE
Percent encoding and `~` are not handled for accounts

### ISSUE THREAD

**Issue by goffi-contrib** [mastodon/mastodon#17222]
### Steps to reproduce the problem

Hello and Happy New Year,

for the context, I'm currently working on a ActivityPub ⟺ XMPP gateway (see https://nlnet.nl/project/Libervia/) and I'm using Mastodon as reference AP implementation.

I need to be able to match Webfinger handle as used in most AP implementations (`username@domain.tld`) to XMPP account and optionally Pubsub node.

According to `acct` URI scheme RFC (https://www.rfc-editor.org/rfc/rfc7565.html#section-7), the username should accept percent encoding and characters such as `~` (which is in the set of [unreserved characters](https://www.rfc-editor.orgrfc/rfc3986#section-2.3).

But Mastodon doesn't accept them (after a quick search in the code, it seems to be limited by `USERNAME_RE` regex at https://github.com/mastodon/mastodon/blob/main/app/models/account.rb#L61), which I believe is a bug. 

**Steps to reproduce:**

1. go to search field
2. enter an account handler using either `~` or percent encoding (e.g. `test~toto@example.org` or `test%40toto@example.org`)
3. see if a HTTP request is done 

Thanks!

### Expected behaviour

HTTP request is done to webfinger service of the targer server to find specified account actor's URL

### Actual behaviour

No request is done and `0 results` is shown

### Specifications

Mastodon from main git branch (at commit c5606db9dae58aec155751105e58d85170efef98) has been used, but it's most probably affecting all versions of Mastodon.

**Comment by ClearlyClaire** [comment]
Indeed, it's a limitation of Mastodon, it only accepts a subset of valid `acct:` URIs (actually, there is one more departure: it is case-insensitive while the spec says the local part is case-sensitive).

I am not yet sure how feasible it would be to widen the range of allowed characters in Mastodon.

**Comment by goffi-contrib** [comment]
Thanks for the quick answer. In my specific use case I'll use a workaround for now, as I don't see a fix largely deployed anytime soon. But it would be far better to see Mastodon (and other AP implementations) handling all the characters specified.

**Comment by MichaelFedora** [comment]
Since #10487 was fairly straightforward, I wonder if changing the

```ruby
USERNAME_RE   = /[a-z0-9_]+([a-z0-9_\.-]+[a-z0-9_]+)?/i
```

to

```ruby
USERNAME_RE   = /\w+(?:[\.\-\~]\w+)*/i
```

would "just work." This would:
- simplify it, since `\w` means `[A-Za-z0-9_]`, and case is already ignored with the tailing `i`
- stop allowing multiple periods/dashes in a row (might be technically allowed but probably okay to be opinionated with)
- make the optional group non-capturing

This would being it (mostly) into [[RFC 3986§2.3](https://www.rfc-editor.org/rfc/rfc3986#section-2.3)] spec, which is

```
unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
```

`ALPHA / DIGIT / "_"` being handled by `\w` and `"." / "-" / "~"` being handled by `[\.\-\~]`.

---

I'm curious as to your workaround @goffi-contrib ? I was considering using tildes as a context/identifier separator for an AP implementation of mine, but since most AP's seem to not enjoy those I guess I'll have to find another separator as well 😅 .

**Comment by MichaelFedora** [comment]
Sure that makes sense. URIs are an odd thing to validate.

Could just be `/[\w\.\-\~]+/` but would allow weird usernames like `_~...@domain.com`. But I guess that's a valid uri?

**Comment by goffi-contrib** [comment]
Sorry for the delayed response; I've been juggling multiple tasks, and this slipped my mind. My main concern is that percent-encoding is not supported when it should be. XMPP allows various characters, and I need to find a way to encode them. Your suggestion doesn't allow for percent-encoding, so I'll have to stick with my current, somewhat cumbersome solution (namely using `.` as a workaround, combined with the use of `___`).

To summarize:

- Your proposal still permits `---` and `___`, which I need for my workaround, so it doesn't break my current implementation, which is a good thing.
- Supporting percent-encoding is the way to go; it aligns with what is defined in the RFC and is the most straightforward option to handle use cases like mine.

**Comment by silverpill** [comment]
+1 for supporting percent encoding

This would be very handy for all kinds of bridges, including translators between incompatible implementations of ActivityPub
